In [1]:
## Emilie Helen Wolf
## December 4, 2021
## Example of creating linear regression model with multiple variables

# import packages for numerical arrays, dataframes, and linear regression
import numpy as np
import pandas as pd
from sklearn import linear_model

In [2]:
# upload the dataframe
df = pd.read_csv('hiring.csv')
df

experience  test_score(out of 10)  interview_score(out of 10)  salary($)
0        NaN                    8.0                           9      50000
1        NaN                    8.0                           6      45000
2       five                    6.0                           7      60000
3        two                   10.0                          10      65000
4      seven                    9.0                           6      70000
5      three                    7.0                          10      62000
6        ten                    NaN                           7      72000
7     eleven                    7.0                           8      80000

In [3]:
# Fill in missing values for experience with 'zero'
df['experience'] = df['experience'].fillna('zero')
df

experience  test_score(out of 10)  interview_score(out of 10)  salary($)
0       zero                    8.0                           9      50000
1       zero                    8.0                           6      45000
2       five                    6.0                           7      60000
3        two                   10.0                          10      65000
4      seven                    9.0                           6      70000
5      three                    7.0                          10      62000
6        ten                    NaN                           7      72000
7     eleven                    7.0                           8      80000

In [4]:
# fill in missing values for test_score with median
df['test_score(out of 10)'] = df['test_score(out of 10)'].fillna(df['test_score(out of 10)'].median())
df

experience  test_score(out of 10)  interview_score(out of 10)  salary($)
0       zero                    8.0                           9      50000
1       zero                    8.0                           6      45000
2       five                    6.0                           7      60000
3        two                   10.0                          10      65000
4      seven                    9.0                           6      70000
5      three                    7.0                          10      62000
6        ten                    8.0                           7      72000
7     eleven                    7.0                           8      80000

In [5]:
# install word2number to quickly convert string to digit
pip install word2number

  Created wheel for word2number: filename=word2number-1.1-py3-none-any.whl size=5586 sha256=c15c303afc8b0f16a8efdf0ac4bc287ff7bfeaea0389bf290f87a5cacd81e6c6
  Stored in directory: c:\users\water\appdata\local\pip\cache\wheels\cb\f3\5a\d88198fdeb46781ddd7e7f2653061af83e7adb2a076d8886d6
Successfully built word2number
Note: you may need to restart the kernel to use updated packages.


In [7]:
from word2number import w2n

In [8]:
# convert words to numbers in experience column
df['experience'] = df['experience'].apply(w2n.word_to_num)
df

experience  test_score(out of 10)  interview_score(out of 10)  salary($)
0           0                    8.0                           9      50000
1           0                    8.0                           6      45000
2           5                    6.0                           7      60000
3           2                   10.0                          10      65000
4           7                    9.0                           6      70000
5           3                    7.0                          10      62000
6          10                    8.0                           7      72000
7          11                    7.0                           8      80000

In [12]:
# Separate independent and dependent variables
X = df.drop(['salary($)'], axis='columns')
Y = df['salary($)']

In [13]:
# Create LinearRegression class object and fit to the variables in the dataset
reg = linear_model.LinearRegression()
reg.fit(X,Y)

LinearRegression()

In [15]:
# slope
reg.coef_

array([2812.95487627, 1845.70596798, 2205.24017467])

In [16]:
# y-intercept
reg.intercept_

17737.263464337666

In [19]:
# make predictions

reg.predict([[2,9,6]])

array([53205.96797671])

An employee with 2 years experience, a 9 on the test, and 6 in the interview is predicted to make $53,206

In [18]:
reg.predict([[12,10,10]])

array([92002.18340611])

An employee with 12 years experience, and scored 10 on both the test and the interview is predicted to earn $92,002